bindings.c

```diff
@@ -4092,6 +4292,363 @@ static int proc_swaps_read(char *buf, size_t size, off_t offset,
+				continue;
+			}
+			strcpy(path_dir, path);
+			strcat(path_dir, "/");
+			strcat(path_dir, file->d_name);
+			pd = depth - 1;
+			sum = calc_pid(pid_buf, path_dir, pd, sum, cfd);
+			free(path_dir);
+		}
+	}
+	closedir(dir);
+
+	/* realloc() failed */
+	if (sum == -1)
+		goto out;
+
+	strcat(path, "/cgroup.procs");
+	fd = openat(cfd, path, O_RDONLY);
+	if (fd < 0)
+		goto out;
+
+	f = fdopen(fd, "r");
+	if (!f)
+		goto out;
+
+	while (getline(&line, &linelen, f) != -1) {
+		pid = (char **)realloc(*pid_buf, sizeof(char *) * (sum + 1));
+		if (pid == NULL) {
+			lxcfs_error("%s\n", "realloc error!");
+			for (; sum > 0; sum--)
+				free((*pid_buf)[sum - 1]);
+			sum = -1;
+			goto out;
+		}
+		*pid_buf = pid;
+		do {
+			*(*pid_buf + sum) = (char *)malloc(strlen(line) + 3);
+		} while (*(*pid_buf + sum) == NULL);
+		strcpy(*(*pid_buf + sum), line);
+		sum++;
+	}
+	fclose(f);
+out:
+	free(path);
+	return sum;
+}
+/*
+ * calc_load calculates the load according to the following formula:
+ * load1 = load0 * exp + active * (1 - exp)
+ *
+ * load1: the new loadavg.
+ * load0: the former loadavg.
+ * active: the total number of running pid at this moment.
+ * exp: the fixed-point defined in the beginning.
+ */
+static unsigned long
+calc_load(unsigned long load, unsigned long exp, unsigned long active)
+{
+	unsigned long newload;
+
+	active = active > 0 ? active * FIXED_1 : 0;
+	newload = load * exp + active * (FIXED_1 - exp);
+	if (active >= load)
+		newload += FIXED_1 - 1;
+
+	return newload / FIXED_1;
+}
+
+/*
+ * Return 0 means that container p->cg is closed.
+ * Return -1 means that error occurred in refresh.
+ * Positive num equals the total number of pid.
+ */
+static int refresh_load(struct load_node *p, char *path)
+{
+	FILE *f = NULL;
+	char **idbuf;
+	char proc_path[256];
+	int i, ret, run_pid = 0, total_pid = 0, last_pid = 0;
+	char *line = NULL;
+	size_t linelen = 0;
+	int sum, length;
+	DIR *dp;
+	struct dirent *file;
+
+	idbuf = (char **)malloc(sizeof(char *));
+	if (!idbuf) {
+		lxcfs_error("Refresh container:%s failed for malloc!\n", p->cg);
+		return -1;
+	}
+	sum = calc_pid(&idbuf, path, DEPTH_DIR, 0, p->cfd);
+	/*  normal exit and abnormal exit */
+	if (sum == 0 || sum == -1)
+		goto out;
+
+	for (i = 0; i < sum; i++) {
+		/*clean up '\n' */
+		length = strlen(idbuf[i])-1;
+		idbuf[i][length] = '\0';
+		do {
+			ret = snprintf(proc_path, 256, "/proc/%s/task", idbuf[i]);
+		} while (ret < 0);
+
+		dp = opendir(proc_path);
+		if (!dp) {
+			lxcfs_error("%s\n", "Open proc_path failed in refresh_load.");
+			continue;
+		}
+		while ((file = readdir(dp)) != NULL) {
+			if (strncmp(file->d_name, ".", 1) == 0)
+				continue;
+			total_pid++;
+			/* We make the biggest pid become last_pid.*/
+			ret = atof(file->d_name);
+			last_pid = (ret > last_pid) ? ret : last_pid;
+
+			do {
+				ret = snprintf(proc_path, 256, "/proc/%s/task/%s/status", idbuf[i], file->d_name);
+			} while (ret < 0);
```

Comment:
Same comment as above:
> Hmm, why? That looks pretty dangerous to be honest. snprintf() isn't guaranteed to allocate any memory internally so the OOM kill won't kill it which means that this could spin indefinitely. If the cause snprintf() error is e.g. EIO I don't see who calling this multiple times would fix anything. I think you should just handle the error once.